pkg/export/gcm/promtest/prometheus.go

```diff
@@ -0,0 +1,179 @@
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	_, port, err := net.SplitHostPort(list.Addr().String())
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	s := http.Server{Handler: m}
+
+	go func() { _ = s.Serve(list) }()
+	env.AddCloser(func() { _ = s.Close() })
+
+	p.p = newPrometheus(env, name, p.image, net.JoinHostPort(env.HostAddr(), port), nil)
+	if err := e2e.StartAndWaitReady(p.p); err != nil {
+		t.Fatalf("can't start Prometheus: %v", err)
+	}
+
+	cl, err := api.NewClient(api.Config{Address: "http://" + p.p.Endpoint("http")})
+	if err != nil {
+		t.Fatalf("create Prometheus client: %s", err)
+	}
+
+	return v1.NewAPI(cl), map[string]string{"job": "test"}
+}
+
+func newPrometheus(env e2e.Environment, name string, image string, scrapeTargetAddress string, flagOverride map[string]string) *e2emon.Prometheus {
+
+	if image == "" {
+		image = "quay.io/prometheus/prometheus:v2.47.2"
+	}
+	ports := map[string]int{"http": 9090}
+
+	f := env.Runnable(name).WithPorts(ports).Future()
+	config := fmt.Sprintf(`
+global:
+  external_labels:
+    collector: %v
+scrape_configs:
+- job_name: 'test'
+  scrape_interval: 5s
+  scrape_timeout: 5s
+  static_configs:
+  - targets: [%s]
+  metric_relabel_configs:
+  - regex: instance
+    action: labeldrop
+`, name, scrapeTargetAddress)
+	if err := os.WriteFile(filepath.Join(f.Dir(), "prometheus.yml"), []byte(config), 0600); err != nil {
+		return &e2emon.Prometheus{Runnable: e2e.NewFailedRunnable(name, fmt.Errorf("create prometheus config failed: %w", err))}
+	}
+
+	args := map[string]string{
+		"--web.listen-address":               fmt.Sprintf(":%d", ports["http"]),
+		"--config.file":                      filepath.Join(f.Dir(), "prometheus.yml"),
+		"--storage.tsdb.path":                f.Dir(),
+		"--enable-feature=exemplar-storage":  "",
+		"--enable-feature=native-histograms": "",
+		"--storage.tsdb.no-lockfile":         "",
+		"--storage.tsdb.retention.time":      "1d",
+		"--storage.tsdb.wal-compression":     "",
+		"--storage.tsdb.min-block-duration":  "2h",
+		"--storage.tsdb.max-block-duration":  "2h",
+		"--web.enable-lifecycle":             "",
+		"--log.format":                       "json",
+		"--log.level":                        "info",
+	}
+	if flagOverride != nil {
+		args = e2e.MergeFlagsWithoutRemovingEmpty(args, flagOverride)
+	}
+
+	p := e2emon.AsInstrumented(f.Init(e2e.StartOptions{
+		Image:     image,
+		Command:   e2e.NewCommandWithoutEntrypoint("prometheus", e2e.BuildArgs(args)...),
+		Readiness: e2e.NewHTTPReadinessProbe("http", "/-/ready", 200, 200),
+		User:      strconv.Itoa(os.Getuid()),
+	}), "http")
+
+	return &e2emon.Prometheus{
+		Runnable:     p,
+		Instrumented: p,
+	}
+}
+
+func (p *promBackend) injectScrapes(t testing.TB, scrapeRecordings [][]*dto.MetricFamily, timeout time.Duration) {
+	t.Helper()
+
+	p.g.mu.Lock()
+	p.g.i = 0
+	p.g.plannedScrapes = scrapeRecordings
+	p.g.mu.Unlock()
+
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
```

Comment:
Why not just pass the ctx into the function args?

Reply:
We could, but that would just move those two lines to the caller - more work for caller essentially. What's the benefit?  (:

Reply:
Just provides more flexibility. I noticed that the k8s client libraries used to do this (pass timeout in arg) but they deprecated those methods and switched to just using ctx (e.g. see their wait package). I guess we can add it if we need that function. Thanks!

Reply:
Yup, luckly we don't need to worry about compatibility here (: